### PR TITLE
Unit tests for BackgroundTaskBuilder API

### PR DESCRIPTION
**Pull request by godlytalias**
UnitTests added for [BackgroundTaskBuilder](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/BackgroundTask/BackgroundTaskBuilder.md) API

- Ensures Task Registration is successful
- Ensures caching of ClsId in ApplicationData works as expected

### CODE REVIEW

File: test/BackgroundTaskTests/BackgroundTaskTests.vcxproj [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    </Reference>
+  </ItemGroup>
+  <ItemGroup>
+    <Xml Include="BackgroundTaskBuilder-AppxManifest.xml">
+      <DeploymentContent>true</DeploymentContent>
+    </Xml>
+  </ItemGroup>
+  <Target Name="CopyFiles" AfterTargets="AfterBuild">
+    <Copy SourceFiles="$(OutDir)\..\WindowsAppRuntime_BootstrapDLL\Microsoft.WindowsAppRuntime.Bootstrap.dll" DestinationFolder="$(OutDir)" />
+    <Copy SourceFiles="$(OutDir)\..\WindowsAppRuntime_UniversalBGTaskDLL\Microsoft.Windows.ApplicationModel.Background.UniversalBGTask.dll" DestinationFolder="$(OutDir)" />
```
Comment (Hemantxk) [inline test/BackgroundTaskTests/BackgroundTaskTests.vcxproj:188]: UniversalBGTask.dll doesn't get copied over from WindowsAppSDKBackgroundTask flag?

**Review by Hemantxk — APPROVED** [review]

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        TEST_METHOD_CLEANUP(MethodUninit)
+        {
+            VERIFY_IS_TRUE(TP::IsPackageRegistered_WindowsAppRuntimeFramework());
+            return true;
+        }
+
+		TEST_METHOD(TestBackgroundTaskRegistration)
+		{
+            if (!::Microsoft::Windows::ApplicationModel::Background::Feature_BackgroundTask::IsEnabled())
```
Comment (hiteshkrmsft) [inline test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:59]: We can add more logs here for better diagnosability

Comment (hiteshkrmsft) [inline test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:59]: we ca put these under a try-catch block to do a better error handling.

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                Log::Result(TestResults::Skipped, L"BackgroundTaskBuilder API Features are not enabled.");
+                return;
+            }
+            winrt::Microsoft::Windows::ApplicationModel::Background::BackgroundTaskBuilder builder;
```
Comment (neerajmsgithub) [inline test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:64]: nit: you already have "using winrt::Windows::ApplicationModel::Background;"  in line 15

**Review by neerajmsgithub — APPROVED** [review]

**Review by hiteshkrmsft — APPROVED** [review]

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            return true;
+        }
+
+        TEST_METHOD_CLEANUP(MethodUninit)
+        {
+            VERIFY_IS_TRUE(TP::IsPackageRegistered_WindowsAppRuntimeFramework());
+            return true;
+        }
+
+		TEST_METHOD(TestBackgroundTaskRegistration)
```
Comment (neerajmsgithub) [inline test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:57]: This only covers registration of background task. Do we need more UTs?

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp (hunk 3 above)
Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:64]: This one is Microsoft.* namespace which is not added to this scope and need explicit directive usage.

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp (hunk 4 above)
Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:57]: Our newly added API takes care only of registration, So just added UT for that for now.

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp (hunk 2 above)
Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:59]: Please clarify on this more, We already have logs for `IsEnabled` check and for other assertion checks there will be logs created as well

Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:59]: Whether the comment here is about `IsEnabled` check? We always follow the `if-else` checks for that

File: test/BackgroundTaskTests/BackgroundTaskTests.vcxproj (hunk 1 above)
Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskTests.vcxproj:188]: Here we don't have the entire WinAppSDK package to do those build tasks

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp (hunk 2 above)
Comment (hiteshkrmsft) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:59]: Agree. But you can consider adding comments using Log::Comment mentioning what is happening in the test. This would give more clarity on test status. But this is optional.

Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp:58]: Test names are quite self explanatory, Will add later if required

File: test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	{
+    public:
+        BEGIN_TEST_CLASS(BackgroundTaskBuilderTests)
+            TEST_CLASS_PROPERTY(L"ThreadingModel", L"MTA")
+            TEST_CLASS_PROPERTY(L"RunFixtureAs:Class", L"RestrictedUser")
+            TEST_CLASS_PROPERTY(L"RunAs", L"UAP")
+            TEST_CLASS_PROPERTY(L"UAP:AppxManifest", L"BackgroundTaskBuilder-AppxManifest.xml")
+            END_TEST_CLASS()
+
+        TEST_CLASS_SETUP(ClassInit)
```
[the remark this answers is not in the document]
Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskBuilderTests.cpp]: Done, Thanks

File: test/BackgroundTaskTests/BackgroundTaskTests.vcxproj
[the remark this answers is not in the document]
Comment (godlytalias) [reply test/BackgroundTaskTests/BackgroundTaskTests.vcxproj:1]: Added, Thanks

File: WindowsAppRuntime.sln [hunk 6]
```diff
@@ -2251,7 +2273,6 @@ Global
 		dev\UndockedRegFreeWinRT\UndockedRegFreeWinRT.vcxitems*{b73ad907-6164-4294-88fb-f3c9c10da1f1}*SharedItemsImports = 4
 		dev\PowerNotifications\PowerNotifications.vcxitems*{b75c1b22-553c-40e4-b38e-6ab4d01fdb9d}*SharedItemsImports = 9
 		dev\DynamicDependency\API\DynamicDependency.vcxitems*{bf055a59-0919-4e34-9b76-dd055495cc5a}*SharedItemsImports = 9
-		test\inc\inc.vcxitems*{c62688a1-16a0-4729-b6ed-842f4faa29f3}*SharedItemsImports = 4
```
Comment (Hemantxk) [inline WindowsAppRuntime.sln:2254]: Is this removal intentional?

Comment (godlytalias) [reply WindowsAppRuntime.sln]: Came up due to some VS changes, reverted it for being on safe side

**Review by akanpatel2206 — APPROVED** [review]
